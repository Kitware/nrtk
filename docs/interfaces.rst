##########
Interfaces
##########

The NRTK API consists of a number of object-oriented functor interfaces for item-response curve
(IRC) generation,
namely for assessing model response to perturbations on given input data. These interfaces focus on black-box IRC
generation. In addition to the driver, or *generator*, of this task, there are two other main components: reference
image perturbation in preparation for black-box testing and black-box scoring for model outputs. The
``PeturbImageFactory`` interface provides the utility to easily vary specified parameters of a particular perturber.
The generator will execute upon a given perturber factory as well as a given model and scorer to generate the IRC for
given input data. We define a few similar interfaces for performing the IRC generation, separated by the intermediate
algorithmic use cases, one for object detection and one for image classification.


We explicitly do not require an abstraction for the black-box operations to fit inside. This is intended to allow for
applications using these interfaces while leveraging existing functionality, which only need to perform data formatting
to fit the input defined here. Note, however, some interfaces are defined for certain black-box concepts as part of the
SMQTK ecosystem (e.g. in `SMQTK-Classifier <https://github.com/Kitware/SMQTK-Classifier>`_,
`SMQTK-Detection
<https://github.com/Kitware/SMQTK-Detection>`_, and other SMQTK-* modules).


These interfaces are based on the plugin and configuration features provided by
`SMQTK-Core <https://github.com/Kitware/SMQTK-Core>`_, to allow convenient hooks into implementation, discoverability,
and factory generation from runtime configuration. This allows for both opaque discovery of interface implementations
from a class-method on the interface class object, as well as instantiation of a concrete instance via a JSON-like
configuration fed in from an outside resource.

.. figure:: figures/api-docs-fig-01.svg

   Figure 1: Abstract Interface Inheritance.

.. When adding new classes within interfaces, sort them alphabetically.

------------------
Image Perturbation
------------------

Interface: PerturbImage
-----------------------
.. autoclass:: nrtk.interfaces.perturb_image.PerturbImage
   :members:
   :special-members:

--------------------
Perturbation Factory
--------------------

Interface: PerturbImageFactory
------------------------------
.. autoclass:: nrtk.interfaces.perturb_image_factory.PerturbImageFactory
   :members:
   :special-members:

-------------
Image Metrics
-------------

Interface: ImageMetric
----------------------
.. autoclass:: nrtk.interfaces.image_metric.ImageMetric
   :members:
   :special-members:

-------
Scoring
-------

Interface: ScoreDetections
--------------------------
.. autoclass:: nrtk.interfaces.score_detections.ScoreDetections
   :members:
   :special-members:

---------------------------------
End-to-End Generation and Scoring
---------------------------------

Interface: GenerateObjectDetectorBlackboxResponse
-------------------------------------------------
.. autoclass:: nrtk.interfaces.gen_object_detector_blackbox_response.GenerateObjectDetectorBlackboxResponse
   :members:
   :special-members:
