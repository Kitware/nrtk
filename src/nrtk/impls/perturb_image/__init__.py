"""Module for all PerturbImage implementations"""
