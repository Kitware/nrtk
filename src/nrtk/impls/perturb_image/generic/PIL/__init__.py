"""Module for PIL based implementations of PerturbImage"""
