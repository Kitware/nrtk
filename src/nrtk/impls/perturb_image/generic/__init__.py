"""Module for all generic implementations of PerturbImage"""
