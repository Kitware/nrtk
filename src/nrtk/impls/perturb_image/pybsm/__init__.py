"""Module for PyBSM based implementations of PerturbImage"""
