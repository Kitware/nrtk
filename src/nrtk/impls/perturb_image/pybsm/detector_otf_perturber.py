from __future__ import annotations

from typing import Any, Dict, Optional, Type, TypeVar

try:
    import cv2

    cv2_available = True
except ImportError:
    cv2_available = False
import numpy as np
import pybsm.radiance as radiance
from pybsm.otf.functional import detector_OTF, otf_to_psf, resample_2D
from pybsm.utils import load_database_atmosphere, load_database_atmosphere_no_interp
from smqtk_core.configuration import (
    from_config_dict,
    make_default_config,
    to_config_dict,
)

from nrtk.impls.perturb_image.pybsm.scenario import PybsmScenario
from nrtk.impls.perturb_image.pybsm.sensor import PybsmSensor
from nrtk.interfaces.perturb_image import PerturbImage

C = TypeVar("C", bound="DetectorOTFPerturber")


class DetectorOTFPerturber(PerturbImage):
    def __init__(
        self,
        sensor: Optional[PybsmSensor] = None,
        scenario: Optional[PybsmScenario] = None,
        w_x: Optional[float] = None,
        w_y: Optional[float] = None,
        f: Optional[float] = None,
        interp: Optional[bool] = True,
    ) -> None:
        """Initializes the DetectorOTFPerturber.

        :param sensor: pyBSM sensor object.
        :param scenario: pyBSM scenario object.
        :param w_x: Detector width in the x direction (m).
        :param w_y: Detector width in the y direction (m).
        :param f: Focal length (m).
        :param interp: a boolean determinings whether load_database_atmosphere is used with or without
                       interpoloation

        If a value is provided for w_x, w_y and/or f that value(s) will be used in
        the otf calculation.

        If both sensor and scenario parameters are provided, but not w_x, w_y and/or f, the
        value(s) of w_x, w_y and/or f will come from the sensor and scenario objects.

        If either sensor or scenario parameters are absent, default values
        will be used for both sensor and scenario parameters (except for w_x/w_y/f, as defined
        below).

        If any of w_x, w_y, or f are absent and sensor/scenario objects are also absent,
        the absent value(s) will default to 4um for w_x/w_y and 50mm for f
        """
        if not self.is_usable():
            raise ImportError("OpenCV not found. Please install 'nrtk[graphics]' or 'nrtk[headless]'.")
        if sensor and scenario:
            if interp:
                atm = load_database_atmosphere(scenario.altitude, scenario.ground_range, scenario.ihaze)
            else:
                atm = load_database_atmosphere_no_interp(scenario.altitude, scenario.ground_range, scenario.ihaze)

            _, _, spectral_weights = radiance.reflectance_to_photoelectrons(
                atm, sensor.create_sensor(), sensor.int_time
            )

            wavelengths = spectral_weights[0]
            weights = spectral_weights[1]

            # cut down the wavelength range to only the regions of interests
            mtf_wavelengths = wavelengths[weights > 0.0]

            D = sensor.D  # noqa: N806

            self.w_x = w_x if w_x is not None else sensor.w_x
            self.w_y = w_y if w_y is not None else sensor.w_y
            self.f = f if f is not None else sensor.f

            self.slant_range = np.sqrt(scenario.altitude**2 + scenario.ground_range**2)
            self.ifov = (sensor.p_x + sensor.p_y) / 2 / self.f
        else:
            self.w_x = w_x if w_x is not None else 4e-6
            self.w_y = w_y if w_y is not None else 4e-6
            self.f = f if f is not None else 50e-3

            # Assume visible spectrum of light
            self.ifov = -1
            self.slant_range = -1
            mtf_wavelengths = np.array([0.58 - 0.08, 0.58 + 0.08]) * 1.0e-6
            # Default value for lens diameter
            D = 0.003  # noqa: N806

        self.sensor = sensor
        self.scenario = scenario
        self.interp = interp

        # Assume if nothing else cuts us off first, diffraction will set the
        # limit for spatial frequency that the imaging system is able
        # to resolve is (1/rad).
        cutoff_frequency = D / np.min(mtf_wavelengths)
        u_rng = np.linspace(-1.0, 1.0, 1501) * cutoff_frequency
        v_rng = np.linspace(1.0, -1.0, 1501) * cutoff_frequency

        # meshgrid of spatial frequencies out to the optics cutoff
        uu, vv = np.meshgrid(u_rng, v_rng)
        self.df = (abs(u_rng[1] - u_rng[0]) + abs(v_rng[0] - v_rng[1])) / 2

        self.det_OTF = detector_OTF(uu, vv, self.w_x, self.w_y, self.f)

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """:raises: ValueError if 'img_gsd' not present in additional_params"""
        if additional_params is None:
            additional_params = dict()

        if self.ifov >= 0 and self.slant_range >= 0:
            if "img_gsd" not in additional_params:
                raise ValueError("'img_gsd' must be present in image metadata for this perturber")

            ref_gsd = additional_params["img_gsd"]
            psf = otf_to_psf(self.det_OTF, self.df, 2 * np.arctan(ref_gsd / 2 / self.slant_range))

            # filter the image
            blur_img = cv2.filter2D(image, -1, psf)

            # resample the image to the camera's ifov
            sim_img = resample_2D(blur_img, ref_gsd / self.slant_range, self.ifov)
        else:
            # Default is to set dxout param to same value as dxin
            psf = otf_to_psf(self.det_OTF, self.df, 1 / (self.det_OTF.shape[0] * self.df))

            sim_img = cv2.filter2D(image, -1, psf)

        return sim_img.astype(np.uint8)

    @classmethod
    def get_default_config(cls) -> Dict[str, Any]:
        cfg = super().get_default_config()
        cfg["sensor"] = make_default_config([PybsmSensor])
        cfg["scenario"] = make_default_config([PybsmScenario])
        return cfg

    @classmethod
    def from_config(cls: Type[C], config_dict: Dict, merge_default: bool = True) -> C:
        config_dict = dict(config_dict)
        sensor = config_dict.get("sensor", None)
        if sensor is not None:
            config_dict["sensor"] = from_config_dict(sensor, [PybsmSensor])
        scenario = config_dict.get("scenario", None)
        if scenario is not None:
            config_dict["scenario"] = from_config_dict(scenario, [PybsmScenario])

        return super().from_config(config_dict, merge_default=merge_default)

    def get_config(self) -> Dict[str, Any]:
        sensor = to_config_dict(self.sensor) if self.sensor else None
        scenario = to_config_dict(self.scenario) if self.scenario else None

        config = {
            "sensor": sensor,
            "scenario": scenario,
            "w_x": self.w_x,
            "w_y": self.w_y,
            "f": self.f,
            "interp": self.interp,
        }

        return config

    @classmethod
    def is_usable(cls) -> bool:
        # Requires opencv to be installed
        return cv2_available
