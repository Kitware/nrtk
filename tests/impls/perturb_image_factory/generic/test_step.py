from __future__ import annotations

import json
from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise
from pathlib import Path
from typing import Any

import numpy as np
import pytest
from PIL import Image
from smqtk_core.configuration import (
    configuration_test_helper,
    from_config_dict,
    to_config_dict,
)
from syrupy.assertion import SnapshotAssertion

from nrtk.impls.perturb_image.pybsm.detector_otf_perturber import DetectorOTFPerturber
from nrtk.impls.perturb_image.pybsm.jitter_otf_perturber import JitterOTFPerturber
from nrtk.impls.perturb_image.pybsm.turbulence_aperture_otf_perturber import TurbulenceApertureOTFPerturber
from nrtk.impls.perturb_image_factory.generic.step import StepPerturbImageFactory
from nrtk.interfaces.perturb_image import PerturbImage
from nrtk.interfaces.perturb_image_factory import PerturbImageFactory
from tests.impls.perturb_image.test_perturber_utils import pybsm_perturber_assertions
from tests.test_utils import CustomFloatSnapshotExtension

DATA_DIR = Path(__file__).parents[3] / "data"
INPUT_IMG_FILE_PATH = "./examples/pybsm/data/M-41 Walker Bulldog (USA) width 319cm height 272cm.tiff"


@pytest.fixture()  # noqa:PT001
def snapshot_custom(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(lambda: CustomFloatSnapshotExtension())


class DummyPerturber(PerturbImage):
    def __init__(self, param_1: int = 1, param_2: int = 2) -> None:
        self.param_1 = param_1
        self.param_2 = param_2

    def perturb(
        self,
        image: np.ndarray,
        _: dict[str, Any] | None = None,
    ) -> np.ndarray:  # pragma: no cover
        return np.copy(image)

    def get_config(self) -> dict[str, Any]:
        return {"param_1": self.param_1, "param_2": self.param_2}


class TestStepPerturbImageFactory:
    @pytest.mark.parametrize(
        ("perturber", "theta_key", "start", "stop", "step", "to_int", "expected"),
        [
            (DummyPerturber, "param_1", 1.0, 6.0, 2.0, True, (1.0, 3.0, 5.0)),
            (DummyPerturber, "param_2", 3.0, 9.0, 3.0, True, (3.0, 6.0)),
            (DummyPerturber, "param_1", 3.0, 9.0, 1.5, False, (3.0, 4.5, 6.0, 7.5)),
            (DummyPerturber, "param_1", 4.0, 4.0, 1.0, False, ()),
        ],
    )
    def test_iteration(
        self,
        perturber: type[PerturbImage],
        theta_key: str,
        start: float,
        stop: float,
        step: float,
        to_int: bool,
        expected: tuple[float, ...],
    ) -> None:
        """Ensure factory can be iterated upon and the varied parameter matches expectations."""
        factory = StepPerturbImageFactory(
            perturber=perturber,
            theta_key=theta_key,
            start=start,
            stop=stop,
            step=step,
            to_int=to_int,
        )
        assert len(expected) == len(factory)
        for idx, p in enumerate(factory):
            assert p.get_config()[theta_key] == expected[idx]

    @pytest.mark.parametrize(
        (
            "perturber",
            "theta_key",
            "start",
            "stop",
            "step",
            "idx",
            "expected_val",
            "expectation",
        ),
        [
            (DummyPerturber, "param_1", 1.0, 6.0, 2.0, 0, 1, does_not_raise()),
            (DummyPerturber, "param_1", 1.0, 6.0, 2.0, 2, 5, does_not_raise()),
            (DummyPerturber, "param_1", 1.0, 6.0, 2.0, 3, -1, pytest.raises(IndexError)),
            (DummyPerturber, "param_1", 1.0, 6.0, 2.0, -1, -1, pytest.raises(IndexError)),
            (DummyPerturber, "param_1", 4.0, 4.0, 1.0, 0, -1, pytest.raises(IndexError)),
        ],
        ids=["first idx", "last idx", "idx == len", "neg idx", "empty iter"],
    )
    def test_indexing(
        self,
        perturber: type[PerturbImage],
        theta_key: str,
        start: float,
        stop: float,
        step: float,
        idx: int,
        expected_val: float,
        expectation: AbstractContextManager,
    ) -> None:
        """Ensure it is possible to access a perturber instance via indexing."""
        factory = StepPerturbImageFactory(perturber=perturber, theta_key=theta_key, start=start, stop=stop, step=step)
        with expectation:
            assert factory[idx].get_config()[theta_key] == expected_val

    @pytest.mark.parametrize(
        ("perturber", "theta_key", "start", "stop", "step", "to_int"),
        [(DummyPerturber, "param_1", 1.0, 5.0, 2.0, False), (DummyPerturber, "param_2", 3.0, 9.0, 3.0, True)],
    )
    def test_configuration(
        self,
        perturber: type[PerturbImage],
        theta_key: str,
        start: float,
        stop: float,
        step: float,
        to_int: bool,
    ) -> None:
        """Test configuration stability."""
        inst = StepPerturbImageFactory(
            perturber=perturber,
            theta_key=theta_key,
            start=start,
            stop=stop,
            step=step,
            to_int=to_int,
        )
        for i in configuration_test_helper(inst):
            assert i.perturber == perturber
            assert i.theta_key == theta_key
            assert i.start == start
            assert i.stop == stop
            assert i.step == step
            assert start in i.thetas
            assert stop not in i.thetas

    @pytest.mark.parametrize(
        ("kwargs", "expectation"),
        [
            (
                {
                    "perturber": DummyPerturber,
                    "theta_key": "param_1",
                    "start": 1.0,
                    "stop": 2.0,
                },
                does_not_raise(),
            ),
            (
                {
                    "perturber": DummyPerturber(1, 2),
                    "theta_key": "param_2",
                    "start": 1.0,
                    "stop": 2.0,
                },
                pytest.raises(TypeError, match=r"Passed a perturber instance, expected type"),
            ),
        ],
    )
    def test_configuration_bounds(self, kwargs: dict[str, Any], expectation: AbstractContextManager) -> None:
        """Test that an exception is properly raised (or not) based on argument value."""
        with expectation:
            StepPerturbImageFactory(**kwargs)

    @pytest.mark.parametrize(
        ("perturber", "theta_key", "start", "stop", "step"),
        [(DummyPerturber, "param_1", 1.0, 5.0, 2.0), (DummyPerturber, "param_2", 3.0, 9.0, 3.0)],
    )
    def test_hydration(
        self,
        tmp_path: Path,
        perturber: type[PerturbImage],
        theta_key: str,
        start: float,
        stop: float,
        step: float,
    ) -> None:
        """Test configuration hydration using from_config_dict."""
        original_factory = StepPerturbImageFactory(
            perturber=perturber,
            theta_key=theta_key,
            start=start,
            stop=stop,
            step=step,
        )

        original_factory_config = original_factory.get_config()

        config_file_path = tmp_path / "config.json"
        with open(str(config_file_path), "w") as f:
            json.dump(to_config_dict(original_factory), f)

        with open(str(config_file_path)) as config_file:
            config = json.load(config_file)
            hydrated_factory = from_config_dict(config, PerturbImageFactory.get_impls())
            hydrated_factory_config = hydrated_factory.get_config()

            assert original_factory_config == hydrated_factory_config

    @pytest.mark.parametrize(
        ("config_file_name", "expectation"),
        [
            (
                "nrtk_brightness_config.json",
                does_not_raise(),
            ),
            (
                "nrtk_bad_step_config.json",
                pytest.raises(ValueError, match=r"not a perturber is not a valid perturber."),
            ),
        ],
    )
    def test_hydration_bounds(self, config_file_name: str, expectation: AbstractContextManager) -> None:
        """Test that an exception is properly raised (or not) based on argument value."""
        with expectation, open(str(DATA_DIR / config_file_name)) as config_file:
            config = json.load(config_file)
            from_config_dict(config, PerturbImageFactory.get_impls())

    @pytest.mark.parametrize(
        ("perturber", "modifying_param", "modifying_val", "theta_key", "start", "stop", "step"),
        [
            (JitterOTFPerturber, "s_y", 0, "s_x", 2e-3, 6e-3, 1e-3),
            (DetectorOTFPerturber, "w_x", 0, "w_y", 3e-3, 9e-3, 1e-3),
            (TurbulenceApertureOTFPerturber, "altitude", 250, "D", 40e-5, 40e-3, 66e-4),
        ],
    )
    def test_perturb_instance_modification(
        self,
        snapshot_custom: SnapshotAssertion,
        perturber: type[PerturbImage],
        modifying_param: str,
        modifying_val: float,
        theta_key: str,
        start: float,
        stop: float,
        step: float,
    ) -> None:
        """Test perturber instance modification for a perturber factory."""
        perturber_factory = StepPerturbImageFactory(
            perturber=perturber,
            theta_key=theta_key,
            start=start,
            stop=stop,
            step=step,
            to_int=False,
        )
        img = np.array(Image.open(INPUT_IMG_FILE_PATH))
        img_md = {"img_gsd": 3.19 / 160.0}
        for perturber in perturber_factory:
            setattr(perturber, modifying_param, modifying_val)
            out_img = pybsm_perturber_assertions(perturb=perturber, image=img, expected=None, additional_params=img_md)
            snapshot_custom.assert_match(out_img)
