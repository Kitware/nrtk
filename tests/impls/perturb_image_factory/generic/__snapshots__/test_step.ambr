# serializer version: 1
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001]
  array([[105, 104, 105, ..., 111, 110, 108],
         [105, 104, 105, ..., 111, 110, 108],
         [105, 104, 105, ..., 111, 110, 108],
         ...,
         [146, 147, 147, ..., 150, 149, 150],
         [146, 147, 147, ..., 150, 149, 150],
         [146, 147, 147, ..., 150, 149, 150]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001].1
  array([[100,  99, 100, ..., 107, 106, 105],
         [100,  99, 100, ..., 107, 106, 105],
         [100,  99, 100, ..., 107, 106, 105],
         ...,
         [147, 147, 147, ..., 150, 150, 150],
         [147, 147, 147, ..., 150, 150, 150],
         [147, 147, 147, ..., 150, 150, 150]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001].2
  array([[ 96,  93,  94, ...,  85,  84,  82],
         [ 96,  93,  94, ...,  85,  84,  82],
         [ 96,  93,  94, ...,  85,  84,  82],
         ...,
         [144, 143, 142, ..., 154, 153, 154],
         [144, 143, 142, ..., 154, 153, 154],
         [144, 143, 142, ..., 154, 153, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001].3
  array([[120, 117, 119, ..., 113, 114, 114],
         [119, 117, 119, ..., 113, 114, 114],
         [119, 117, 119, ..., 114, 115, 114],
         ...,
         [113, 111, 110, ..., 153, 152, 152],
         [113, 111, 110, ..., 154, 153, 153],
         [113, 111, 110, ..., 154, 153, 153]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001].4
  array([[255, 255, 255, ..., 246, 245, 250],
         [255, 255, 255, ..., 246, 246, 250],
         [255, 255, 255, ..., 246, 246, 250],
         ...,
         [  0,   0,   0, ...,  50,  48,  45],
         [  0,   0,   0, ...,  50,  48,  45],
         [  0,   0,   0, ...,  50,  48,  44]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[DetectorOTFPerturber-w_x-0-w_y-0.003-0.009-0.001].5
  array([[207, 209, 208, ..., 200, 200, 202],
         [207, 209, 208, ..., 200, 200, 202],
         [207, 209, 208, ..., 200, 200, 202],
         ...,
         [ 39,  37,  40, ...,  62,  60,  57],
         [ 39,  37,  40, ...,  62,  60,  57],
         [ 39,  37,  40, ...,  62,  60,  58]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[JitterOTFPerturber-s_y-0-s_x-0.002-0.006-0.001]
  array([[119, 119, 119, ..., 121, 121, 121],
         [118, 118, 118, ..., 115, 115, 115],
         [120, 120, 120, ..., 126, 127, 127],
         ...,
         [146, 146, 146, ..., 152, 152, 152],
         [147, 147, 147, ..., 151, 151, 151],
         [145, 145, 145, ..., 149, 149, 149]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[JitterOTFPerturber-s_y-0-s_x-0.002-0.006-0.001].1
  array([[118, 118, 118, ..., 121, 121, 121],
         [121, 121, 121, ..., 114, 114, 114],
         [122, 122, 122, ..., 121, 121, 121],
         ...,
         [145, 145, 145, ..., 152, 152, 152],
         [147, 147, 147, ..., 151, 151, 151],
         [144, 144, 144, ..., 149, 149, 149]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[JitterOTFPerturber-s_y-0-s_x-0.002-0.006-0.001].2
  array([[118, 118, 118, ..., 122, 122, 122],
         [122, 122, 122, ..., 113, 113, 113],
         [124, 124, 124, ..., 118, 118, 118],
         ...,
         [143, 143, 143, ..., 152, 152, 152],
         [144, 144, 144, ..., 151, 151, 151],
         [142, 142, 142, ..., 150, 150, 150]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[JitterOTFPerturber-s_y-0-s_x-0.002-0.006-0.001].3
  array([[119, 119, 119, ..., 123, 123, 123],
         [124, 124, 124, ..., 111, 111, 111],
         [125, 125, 125, ..., 115, 115, 115],
         ...,
         [139, 139, 139, ..., 153, 153, 153],
         [140, 140, 140, ..., 152, 152, 152],
         [139, 139, 139, ..., 150, 150, 150]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066]
  array([[141, 145, 143, ..., 133, 137, 137],
         [111, 116, 124, ..., 130, 123, 123],
         [123, 126, 130, ..., 152, 152, 147],
         ...,
         [135, 152, 143, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].1
  array([[141, 145, 143, ..., 133, 137, 137],
         [111, 116, 124, ..., 130, 123, 123],
         [123, 126, 130, ..., 152, 152, 147],
         ...,
         [135, 152, 143, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].2
  array([[141, 145, 143, ..., 133, 137, 137],
         [111, 116, 124, ..., 130, 123, 123],
         [123, 126, 130, ..., 152, 152, 147],
         ...,
         [135, 152, 143, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].3
  array([[140, 144, 143, ..., 133, 137, 137],
         [112, 116, 124, ..., 130, 124, 123],
         [123, 126, 130, ..., 152, 151, 147],
         ...,
         [136, 152, 143, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].4
  array([[140, 144, 142, ..., 133, 136, 137],
         [112, 117, 124, ..., 130, 124, 124],
         [123, 126, 130, ..., 151, 151, 146],
         ...,
         [136, 152, 144, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].5
  array([[140, 144, 142, ..., 133, 136, 136],
         [112, 117, 124, ..., 131, 124, 124],
         [123, 126, 130, ..., 151, 151, 146],
         ...,
         [136, 151, 144, ..., 164, 154, 152],
         [150, 156, 152, ..., 159, 161, 156],
         [156, 159, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
# name: TestStepPerturbImageFactory.test_perturb_instance_modification[TurbulenceApertureOTFPerturber-altitude-250-D-0.0004-0.04-0.0066].6
  array([[139, 143, 142, ..., 133, 136, 136],
         [113, 117, 125, ..., 131, 125, 124],
         [123, 126, 130, ..., 151, 150, 146],
         ...,
         [137, 151, 144, ..., 163, 154, 152],
         [150, 156, 152, ..., 159, 160, 156],
         [156, 158, 152, ..., 159, 154, 154]], dtype=uint8)
# ---
